backend/remote-state/pg/client.go

```diff
@@ -0,0 +1,181 @@
+package pg
+
+import (
+	"context"
+	"crypto/md5"
+	"database/sql"
+	"fmt"
+
+	uuid "github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/terraform/state"
+	"github.com/hashicorp/terraform/state/remote"
+	_ "github.com/lib/pq"
+)
+
+// RemoteClient is a remote client that stores data in a Postgres database
+type RemoteClient struct {
+	Client     *sql.DB
+	Name       string
+	SchemaName string
+
+	// In-flight database transaction. Empty unless Locked.
+	txn  *sql.Tx
+	info *state.LockInfo
+}
+
+func (c *RemoteClient) Get() (*remote.Payload, error) {
+	query := `SELECT data FROM %s.%s WHERE name = $1`
+	var row *sql.Row
+	// Use the open transaction when present
+	if c.txn != nil {
+		row = c.txn.QueryRow(fmt.Sprintf(query, c.SchemaName, statesTableName), c.Name)
+	} else {
+		row = c.Client.QueryRow(fmt.Sprintf(query, c.SchemaName, statesTableName), c.Name)
+	}
+	var data []byte
+	err := row.Scan(&data)
+	switch {
+	case err == sql.ErrNoRows:
+		// No existing state returns empty.
+		return nil, nil
+	case err != nil:
+		return nil, err
+	default:
+		md5 := md5.Sum(data)
+		return &remote.Payload{
+			Data: data,
+			MD5:  md5[:],
+		}, nil
+	}
+}
+
+func (c *RemoteClient) Put(data []byte) error {
+	query := `INSERT INTO %s.%s (name, data) VALUES ($1, $2)
+		ON CONFLICT (name) DO UPDATE
+		SET data = $2 WHERE %s.name = $1`
+	var err error
+	// Use the open transaction when present
+	if c.txn != nil {
+		_, err = c.txn.Exec(fmt.Sprintf(query, c.SchemaName, statesTableName, statesTableName), c.Name, data)
+	} else {
+		_, err = c.Client.Exec(fmt.Sprintf(query, c.SchemaName, statesTableName, statesTableName), c.Name, data)
+	}
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *RemoteClient) Delete() error {
+	query := `DELETE FROM %s.%s WHERE name = $1`
+	var err error
+	// Use the open transaction when present
+	if c.txn != nil {
+		_, err = c.txn.Exec(fmt.Sprintf(query, c.SchemaName, statesTableName), c.Name)
+	} else {
+		_, err = c.Client.Exec(fmt.Sprintf(query, c.SchemaName, statesTableName), c.Name)
+	}
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *RemoteClient) Lock(info *state.LockInfo) (string, error) {
+	var err error
+	var lockID string
+	var txn *sql.Tx
+
+	if info.ID == "" {
+		lockID, err = uuid.GenerateUUID()
+		if err != nil {
+			return "", err
+		}
+		info.Operation = "client"
+		info.ID = lockID
+	}
+
+	if c.txn == nil {
+		// Most strict transaction isolation to prevent cross-talk
+		// between incomplete state transactions.
+		txn, err = c.Client.BeginTx(context.Background(), &sql.TxOptions{
+			Isolation: sql.LevelSerializable,
+		})
+		if err != nil {
+			return "", err
```

Comment:
If there is an error in this function, we want to try and wrap it in a`*state.LockError` with the lock info. The ID isn't as important here if these are canceled if a client disconnects, but the info may be useful to other clients to indicate where the lock is being taken.

Reply:
I see what you mean and [found](https://github.com/hashicorp/terraform/blob/master/state/state.go#L39) that the `LockError` type makes it be retried.

I've revised to return the `*state.LockError` type when retrying `Lock()` makes sense.